### DISCUSSION TITLE
Some unknown javascript error on deployed pages only which I am going to figure out soon

### DISCUSSION THREAD

**Discussion by McDic** [squidfunk/mkdocs-material#6644] [Q&A] [answered]
When the website is loaded, javascript stuffs are not working at all, unlike local.

- Search bar
- Secondary navigation sidebar
- Light/Dark mode switch button

```js
Uncaught TypeError: Cannot read properties of undefined (reading 'getAttribute')
    at index.ts:100:26
    at map.ts:58:33
    at c._next (OperatorSubscriber.ts:70:13)
    at t.next (Subscriber.ts:75:12)
    at mergeInternals.ts:85:24
    at c._next (OperatorSubscriber.ts:70:13)
    at t.next (Subscriber.ts:75:12)
    at e._subscribe (innerFrom.ts:78:18)
    at e._trySubscribe (Observable.ts:244:19)
    at Observable.ts:234:18
```

And this is the section where error occurred.

```ts
/**
 * Watch color palette
 *
 * @param inputs - Color palette element
 *
 * @returns Color palette observable
 */
export function watchPalette(
  inputs: HTMLInputElement[]
): Observable<Palette> {
  const current = __md_get<Palette>("__palette") || {
    index: inputs.findIndex(input => matchMedia(
      input.getAttribute("data-md-color-media")!
    ).matches)
  }

  /* Emit changes in color palette */
  return of(...inputs)
    .pipe(
      mergeMap(input => fromEvent(input, "change").pipe(map(() => input))),
      startWith(inputs[Math.max(0, current.index)]),
      map(input => ({
        index: inputs.indexOf(input),
        color: {
          media:   input.getAttribute("data-md-color-media"),  /* <-------------------------- HERE */
          scheme:  input.getAttribute("data-md-color-scheme"),
          primary: input.getAttribute("data-md-color-primary"),
          accent:  input.getAttribute("data-md-color-accent")
        }
      } as Palette)),
      shareReplay(1)
    )
}
```

... while local `mkdocs serve` is perfectly working without any problem on browser developer mode console. You can visit [this page here](https://blog.mcdic.net/format).

![스크린샷 2024-01-14 121414](https://github.com/squidfunk/mkdocs-material/assets/19789019/de0cd058-e099-4040-963d-84a7f26740b4)

---

Here are `extra_javascripts` and `extra_css` settings in [`mkdocs.yml`](https://github.com/McDic/BlogV2/blob/master/mkdocs.yml).

```yml
extra_javascript:
  - javascripts/mathjax.js
  - javascripts/viewcount.js
  - https://polyfill.io/v3/polyfill.min.js?features=es6
  - https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js

extra_css:
  - stylesheets/admonitions.css
  - stylesheets/extra.css
```

Here is the `libs` block I overrided in [`overrides/main.html`](https://github.com/McDic/BlogV2/blob/master/overrides/main.html).

```html
{% block libs %}

    {{ super() }}
    <script type="text/javascript" src="/javascripts/viewcount.js"></script>

{% endblock %}
```

I use [`mkdocs gh-deploy -v -r origin -b deploy -s --force --no-history`](https://github.com/McDic/BlogV2/blob/master/.github/workflows/deploy.yml) to deploy the blog on Github Pages. This is made by help from the [official documentation of `mkdocs-material`](https://squidfunk.github.io/mkdocs-material/publishing-your-site/?h=github#with-github-actions).

---

I noticed this bug is happening from somewhere of the commit and I wanted to report this, however I need to make a minimal reproduction of this. Maybe I have to make some another github pages for this.

What I am currently guessing is because I removed one palette (automatic mode) to make from 3 to 2 palettes and there is an error on cache or somewhat.

I am opening this discussion in case other people facing similar issues or already know a solution.

**Comment by kamilkrzyskow** [comment, marked as answer]
Hello @McDic,
since the `viewcount.js` is included in the `extra_javascript` there is no need to override `main.html` with the `libs`.
As for the error I don't get it when I visit https://blog.mcdic.net/ so it's likely, as you said, that the palette `index` stored in Local Storage of your browser might be invalid and therefore results in a `undefined`.

**Reply by kamilkrzyskow** [reply]
A possible temporary fix, to make sure other users have a functional website when visiting your blog, would be to add a `<script>` tag to the page, which runs before the bundle.js, and it should contain some sort of validation / fixup of the value within the Local Storage.

**Reply by McDic** [reply]
You were correct, I thought I removed all the cache by removing cookies but it wasn't. I completely wiped out all the browser data and now my browser is working good.

Ok, thanks for letting me know, I am going to make a commit to remove `viewcount.js` thing from overrided `libs` block.

Yeah, it might be good to add some validation script. But to do that I guess I need to learn some more stuffs.

Anyway thanks for your helpful answer.

**Reply by kamilkrzyskow** [reply]
Oof the poor browser data, I wouldn't be so fast to delete it, I hope you didn't lose anything important. 
You can open the browser developer tools and look at, and remove, the Local Storage for each page separately:

<details>
<summary> Screenshot </summary>

![image](https://github.com/squidfunk/mkdocs-material/assets/34622465/4fd75f53-a6a8-40c9-ad23-921fad8b995d)

</details>

You can see the code that loads the selected palette upon website render here:
https://github.com/squidfunk/mkdocs-material/blob/ab5ccbe6b8c49dbf58a25eb066a98b243a382967/src/templates/partials/javascripts/palette.html#L24-L47

Also please, if you have the time, prepare a minimal example and report an issue:
https://squidfunk.github.io/mkdocs-material/guides/creating-a-reproduction/
https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/

**Reply by McDic** [reply]
Ah I see that's the proper way to remove local storage cache. Fortunately the only important stuffs are probably rarely used Google account cookies or something similar to that. So no worries.

Ok, I made a [repository to reproduce](https://github.com/McDic/mkdocs-material-palette-bug-report-2024-01/tree/master) and I am going to make an issue now. Thanks!

**Reply by McDic** [reply]
Oh I see `mkdocs-material` is [not encouraging](https://squidfunk.github.io/mkdocs-material/contributing/reporting-a-bug/) to make a reproducing repository. I guess I am going to make two `.zip` files.

**Reply by McDic** [reply]
Reported https://github.com/squidfunk/mkdocs-material/issues/6645, thanks!

**Reply by kamilkrzyskow** [reply]
~~I think that one zip file should be enough with the steps to reproduce the error, as I'm assuming that changing the 3 palette toggle to the  2 palette toggle caused the crash. Also please use the `info` plugin to generate the zip, not really important in this case as the error comes from JavaScript, but in general this is the procedure.~~

**Reply by kamilkrzyskow** [reply]
> Reported #6645, thanks!

Thank you for being so thorough. 👍
